swift/apple/FirezoneKit/Sources/FirezoneKit/Helpers/AppLogger.swift

```diff
@@ -0,0 +1,326 @@
+      )
+
+      if diskLog == nil {
+        self.logger.error(
+          "LogWriter.performInitialSetup: Unable to switch log. Log entries will not be written to disk."
+        )
+      }
+
+      self.workQueue.async {
+        if case .memory(let memoryLog) = self.logDestination {
+          let bytesWritten =
+            Self.writeMemoryLogToDisk(memoryLog: memoryLog, diskLog: diskLog, logger: self.logger)
+          self.fileSizeAddendum = UInt64(bytesWritten)
+          self.logDestination = .disk(diskLog)
+          self.logger.error(
+            "LogWriter.performInitialSetup: Switched to disk log."
+          )
+        }
+      }
+    }
+  }
+
+  private static func openDiskLog(
+    folderURL: URL, logFileNameBase: String,
+    logIndex: Int, shouldRemoveExistingFile: Bool, logger: Logger
+  ) -> DiskLog? {
+
+    let logFileURL = folderURL.appendingPathComponent(
+      "\(logFileNameBase)\(logIndex).\(Self.logFileNameExtension)")
+    let logFilePath = logFileURL.path
+
+    let fileManager = FileManager.default
+
+    if shouldRemoveExistingFile {
+      do {
+        logger.error("LogWriter.openDiskLog: Removing file at '\(logFilePath, privacy: .public)'")
+        try fileManager.removeItem(atPath: logFilePath)
+      } catch {
+        logger.error(
+          "LogWriter.openDiskLog: Error removing file '\(logFilePath, privacy: .public)'"
+        )
+      }
+    }
+
+    if !shouldRemoveExistingFile && fileManager.fileExists(atPath: logFilePath) {
+      logger.error("LogWriter.openDiskLog: File exists at '\(logFilePath, privacy: .public)'")
+      var fileSize: UInt64? = nil
+      do {
+        let attr = try fileManager.attributesOfItem(atPath: logFilePath)
+        fileSize = attr[FileAttributeKey.size] as? UInt64
+      } catch {
+        logger.error(
+          "LogWriter.openDiskLog: Error getting file attributes of '\(logFilePath, privacy: .public)': \(error, privacy: .public)"
+        )
+        return nil
+      }
+      if let fileSize = fileSize {
+        if let filePointer = fopen(logFilePath, "a") {
+          return DiskLog(logIndex: logIndex, filePointer: filePointer, fileSizeAtOpen: fileSize)
+        } else {
+          logger.error(
+            "LogWriter.openDiskLog: Can't open file '\(logFilePath, privacy: .public)' for appending"
+          )
+          return nil
+        }
+      } else {
+        logger.error(
+          "LogWriter.openDiskLog: Can't figure out file size for '\(logFilePath, privacy: .public)'"
+        )
+        return nil
+      }
+    } else {
+      logger.error("LogWriter.openDiskLog: Creating file at '\(logFilePath, privacy: .public)'")
+      if let filePointer = fopen(logFilePath, "w") {
+        return DiskLog(logIndex: logIndex, filePointer: filePointer, fileSizeAtOpen: 0)
+      } else {
+        logger.error(
+          "LogWriter.openDiskLog: Can't open file '\(logFilePath, privacy: .public)' for writing"
+        )
+        return nil
+      }
+    }
```

Comment:
I think all these error conditions should be critical errors -- can we instead show this to the user via a UI popup when they launch the app or is that too cumbersome?

We probably want these edge cases to be noisy because they're likely to cause other issues and we'd like to hear from users when they occur.